runtime/autoload/provider/clipboard.vim

```diff
@@ -4,43 +4,88 @@
 let s:copy = {}
 let s:paste = {}
 
+" when caching is enabled, store the jobid of the xclip/xsel process
+" keeping ownership of the selection, so we know how long the cache is valid
+let s:selection = { 'owner': 0, 'data': [] }
+
+function! s:selection.on_exit(jobid, data, event)
+  " At this point this nvim instance might already have launched
+  " a new provider instance. Don't drop ownership in this case.
+  if self.owner == a:jobid
+    let self.owner = 0
+  endif
+endfunction
+
+let s:selections = { '*': s:selection, '+': copy(s:selection)}
+
 function! s:try_cmd(cmd, ...)
-  let out = a:0 ? systemlist(a:cmd, a:1, 1) : systemlist(a:cmd, [''], 1)
+  let argv = split(a:cmd, " ")
+  let out = a:0 ? systemlist(argv, a:1, 1) : systemlist(argv, [''], 1)
   if v:shell_error
-    echo "clipboard: error: ".(len(out) ? out[0] : '')
-    return ''
+    echohl WarningMsg
+      echo "clipboard: error: ".(len(out) ? out[0] : '')
+    echohl None
+    return 0
   endif
   return out
 endfunction
 
+let s:cache_enabled = 1
 if executable('pbcopy')
   let s:copy['+'] = 'pbcopy'
   let s:paste['+'] = 'pbpaste'
   let s:copy['*'] = s:copy['+']
   let s:paste['*'] = s:paste['+']
+  let s:cache_enabled = 0
 elseif executable('xclip')
-  let s:copy['+'] = 'xclip -i -selection clipboard'
+  let s:copy['+'] = 'xclip -quiet -i -selection clipboard'
   let s:paste['+'] = 'xclip -o -selection clipboard'
-  let s:copy['*'] = 'xclip -i -selection primary'
+  let s:copy['*'] = 'xclip -quiet -i -selection primary'
   let s:paste['*'] = 'xclip -o -selection primary'
 elseif executable('xsel')
-  let s:copy['+'] = 'xsel -i -b'
+  let s:copy['+'] = 'xsel --nodetach -i -b'
   let s:paste['+'] = 'xsel -o -b'
-  let s:copy['*'] = 'xsel -i -p'
+  let s:copy['*'] = 'xsel --nodetach -i -p'
   let s:paste['*'] = 'xsel -o -p'
 else
   echom 'clipboard: No shell command for communicating with the clipboard found.'
+  echom 'clipboard: see :help nvim-clipboard'
```

Comment:
EXX are for *errors*. It is inconsistent to do this for :echomsg. Also note that

1. These messages will not be translated. Translators currently have reasons to expect all messages mentioned in .po files. Users would expect that all messages will *eventually* be tranlated.
2. When defining new messages searching numbers in runtime/ files would be unexpected from developers that got used to Vim.
3. When merging patches from Vim it may appear that our error message code conflicts with Vim error message code.
4. Exxx messages are expected to be caught as `Vim({command}):E{number}:`, but this is not possible here.
5. It is generally harder to maintain unambiguous error numbers, identifiers like `E:[{plugin}/]{component}/{id}` (e.g. `E:clipboard/nocmd` are much less likely to clash (`E:` is needed to mark error as error in \:help, without making user remember a convention and is thus not much useful: most of time `:h clipboard/nocmd` will normally get you to the same topic as `:h E:clipboard/nocmd`).

I do not agree about \:h in error messages *in the current state*: they either need proper identifiers like described above or explicit \:h: I would remember the topic only if it is either always there (i.e. as error identifier or explicit \:h reference) or if NeoVim often will not work with such messages.